internal/csi-addons/rbd/network_fence.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2021 The Ceph-CSI Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	"github.com/csi-addons/spec/lib/go/fence"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// FenceControllerServer struct of rbd CSI driver with supported methods
+// of CSI-addons networkfence controller service spec.
+type FenceControllerServer struct {
+	*fence.UnimplementedFenceControllerServer
+}
+
+// NewFenceControllerServer creates a new IdentityServer which handles
+// the Identity Service requests from the CSI-Addons specification.
+func NewFenceControllerServer() *FenceControllerServer {
+	return &FenceControllerServer{}
+}
+
+func (fcs *FenceControllerServer) RegisterService(server grpc.ServiceRegistrar) {
+	fence.RegisterFenceControllerServer(server, fcs)
+}
+
+// validateFenceClusterNetworkReq checks the sanity of FenceClusterNetworkRequest.
+func validateNetworkFenceReq(fenceClients []*fence.CIDR, options map[string]string) error {
+	if len(fenceClients) == 0 {
+		return status.Error(codes.InvalidArgument, "cidr block cannot be empty")
+	}
+
+	if value, ok := options["clusterID"]; !ok || value == "" {
+		return status.Error(codes.InvalidArgument, "missing or empty clusterID")
+	}
+
+	return nil
+}
+
+// genNetworkFenceRequest returns a networkFence struct object from the Network fence/unfence request.
+func genNetworkFenceRequest(
+	ctx context.Context, cidrs []*fence.CIDR, fenceOptions map[string]string) (*util.NetworkFence, error) {
+	var err error
+	nwFence := &util.NetworkFence{}
+
+	nwFence.Cidr, err = util.GetCIDR(cidrs)
+	if err != nil {
+		return nil, err
```

Comment:
Descriptive errors are very helpful for debugging. Not always needed for local helpers in the same package, but should be added for calling functions from other packages. Consider this comment for all `return nil, err` lines.

Reply:
Thanks for the suggestion! Updated now